crypto_sign/falcon-1024/clean/Makefile.Microsoft_nmake

```diff
@@ -0,0 +1,23 @@
+# This Makefile can be used with Microsoft Visual Studio's nmake using the command:
+#    nmake /f Makefile.Microsoft_nmake
+
+LIBRARY=libfalcon-1024_clean.lib
+OBJECTS=codec.obj common.obj fft.obj fpr.obj keygen.obj pqclean.obj rng.obj sign.obj vrfy.obj
+
+# Warning C4146 is raised when a unary minus operator is applied to an
+# unsigned type; this has nonetheless been standard and portable for as
+# long as there has been a C standard, and we do that a lot, especially
+# for constant-time computations. Thus, we disable that spurious warning.
```

Comment:
I'm not entirely sure about this, but I also don't exactly know how trivial this warning is. As it worked everywhere, I'm inclined to leave this be.

Reply:
There is some historical context in [Microsoft docs](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4146?view=vs-2019): the main reason for this warning is for people writing "-2147483648" and expecting a signed integer; instead, the constant (without the "-") is too large to fit an "int", and instead has type "unsigned", and the mathematical result of the negation will be an "unsigned" of value +2147483648 instead.

Now, the Microsoft explanation is actually wrong: in C99, or even in C90 on 64-bit systems, "2147483648" as a literal constant will not fit in an "int", so it will be promoted to "long" or "long long" (whichever is large enough), so still signed. Only in C90 (no "long long") on a 32-bit system (so a 32-bit "long") will it become "unsigned". But, at least, this explains what the Microsoft engineers were thinking about.

In many places in the Falcon code, I use negation on unsigned types to get the standard effect, which involves reduction modulo the size of the type (this is specified in the C standard, section 6.2.5/9: "A computation involving unsigned operands can never overflow, because a result that cannot be represented by the resulting unsigned integer type is reduced modulo the number that is one greater than the largest value that can be represented by the resulting type."). Most of the time, this is to transform a bit result (value 0 or 1) into a full-width mask (0x00000000 or 0xFFFFFFFF).

To avoid the C4146 warning, I would have to replace all cases of `-x` with `0-x`, which is cumbersome (it would require extra parentheses) and unsatisfying: I want the compiler to use the `neg` opcode, not a `sub`, and it will do that only if it notices that the subtraction is really a negation in disguise; but, to avoid the warning, I need the compiler _not_ to notice it.

I usually disable the warning with a `#pragma`:

    #if defined _MSC_VER && _MSC_VER
    #pragma warning( disable : 4146 )
    #endif

but then, there will be a `#if` and it will fail `test_preprocessor.py`. Hence the extra flag for the compiler.

Reply:
Great explanation.

The Microsoft explanation probably stems from the fact that they don't support C90, and C99 also not really nicely (only the parts needed to support C++ versions).